frontend/src/app/ParamIdValidator.tsx

```diff
@@ -0,0 +1,24 @@
+import { useParams } from 'react-router-dom'
+
+import { MONGODB_ID_REGEX } from '~constants/routes'
+
+import NotFoundErrorPage from '~pages/NotFoundError'
+
+interface ParamIdValidatorProps {
+  element: React.ReactElement
+}
+
+export const ParamIdValidator = ({ element }: ParamIdValidatorProps) => {
+  const { formId, submissionId, paymentId } = useParams()
+
+  // Bootstrap route validation as suggested by React Router Docs:
+  // https://reactrouter.com/en/main/start/faq#what-happened-to-regexp-routes-paths
+  if (
+    (formId && !formId.match(MONGODB_ID_REGEX)) ||
```

Comment:
nit: Extract match into a function for better readability and maintainability

`const isValidMongoId = (id: string) => id && id.match(MONGODB_ID_REGEX)`


` if (!isValidMongoId(formId) || !isValidMongoId(submissionId) || !isValidMongoId(paymentId)) {
    return <NotFoundErrorPage />
  }`

Reply:
Thanks! Done in [1c97c99](https://github.com/opengovsg/FormSG/pull/6561/commits/1c97c99a2031a5402cd201375a0483f2cd80dc1d)